### ISSUE TITLE
MergeBamAlignment needlessly requires input BAMs to be sorted by queryname

### ISSUE THREAD

**Issue by freeseek** [broadinstitute/gatk#7398]
Generate a toy fastq dataset:
```
(echo "@NB500989:333:HKYJNAFX2:1:11101:24447:1024"
echo "CNGAAAGAAACTAGAGGGGGCCAGGTGCAGTAGCTCACACCTGTAATCTTAGCACTTTGGGAGGCCAAGGCGGGTGGATCATCTGAGGTCAGGAGTTCAAGACCAGCCTGGCCAACATGTTGAAACCCCGTCTCTATTAAAAATACAAAA"
echo "+"
echo "6#A/AEEEEEEEAEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEE<EEEEEEEEAEEEEEEEEEEEEEAEEEAEEEAEEEEEE/EEEEEEEEE/E/E<AEEEEE/6A/<<<EE/AEE/E"
echo "@NB500989:333:HKYJNAFX2:1:11101:10000:1915"
echo "AGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAATCTTGAGTCTGGCAGCAATTTGTAAAGGCTCATAAAAATCTGAAGCTTACATTTTTTGTCAAGTTACCGATGCTTGTGTCT"
echo "+"
echo "AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEAEEEEEAEEEEEEEEEAEEEEEEEEEAEEEEEEEEEEEAAAEAAAEEE/A<AEEEEEEEEEEAA") | \
  gzip > R1.fastq.gz

(echo "@NB500989:333:HKYJNAFX2:1:11101:24447:1024"
echo "NNNNTTGTATTTTTAATAGAGACGGGGTTTCAACATGTTGGCCAGGCTGGTCTTGAACTCCTGACCTCAGATGATCCACCCGCCTTGGCCTCCCAAAGTGCTAAGATTACAGGTGTGAGCTACTGCACCTGGCCCCCTCTAGTTTCTTTC"
echo "+"
echo "####AEEEEEEEEEEEEEEEE/EEEEEEEEEE6EEEEEEEEEEEEA<EEEEEEEEE/<EEEEEEE/EEEEEEEEEEE/EEEEEAEEEEEAEAEEEE/EEEEE/EEEEEEEEEEEEEAEEEEEEEEAE/<AAAA/<A<EE<<EA<A/AAE<"
echo "@NB500989:333:HKYJNAFX2:1:11101:10000:1915"
echo "TTACAGGATCTGAAGAGAGGGAAAAATAAACATGCACGATTATTTAATTCTTTTGGAAAAACTGCATGTAAGTGAAGTTCTCTTTCACAAGACACAAGCATCGGTAACTTGACAAAAAATGTAAGCTTCAGATTTTTATGAGCCTTTACA"
echo "+"
echo "AAAAAEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE/EEEEEEEEEE6EEEEEEEEEEEEAEEEEEE/EEEAEEEEEEEEEEEEEAEEAE/EEEEEEEEEAEEEEEEA<EEEEEE<AAEEEEEEEEEEAEEEEEAEAAEAE/AAEAEAA<E") | \
  gzip > R2.fastq.gz
```

Install bwa and GATK resources:
```
sudo apt install java11-runtime bwa samtools
wget https://github.com/broadinstitute/gatk/releases/download/4.2.1.0/gatk-4.2.1.0.zip
unzip gatk-4.2.1.0.zip
echo dict fasta fasta.fai fasta.64.alt fasta.64.sa fasta.64.amb fasta.64.bwt fasta.64.ann fasta.64.pac | \
  tr ' ' '\n' | xargs -i echo gs://gcp-public-data--broad-references/hg38/v0/Homo_sapiens_assembly38.{} | \
  gsutil -m cp -I .
```

Run FastqToSam with the `-SORT_ORDER "unsorted"` option:
```
gatk-4.2.1.0/gatk \
  FastqToSam \
  -FASTQ R1.fastq.gz \
  -FASTQ2 R2.fastq.gz \
  -OUTPUT unmapped.bam \
  -SAMPLE_NAME SM \
  -SORT_ORDER "unsorted"
```
Notice the option `-SORT_ORDER "unsorted"` which prevents the tool from resorting the reads which can add both computational and storage requirements.

Aligned the data with bwa:
```
gatk-4.2.1.0/gatk \
  SamToFastq \
  -INPUT unmapped.bam \
  -FASTQ /dev/stdout \
  -INTERLEAVE true | \
  bwa mem -K 100000000 -p -v 3 -t 16 -Y Homo_sapiens_assembly38.fasta /dev/stdin | \
  samtools view -1 - > aligned.unmerged.bam
```

Merge unmapped and aligned BAMs:
```
gatk-4.2.1.0/gatk \
  MergeBamAlignment \
  -ALIGNED_BAM aligned.unmerged.bam \
  -UNMAPPED_BAM unmapped.bam \
  -OUTPUT aligned.unsorted.bam \
  -SORT_ORDER "unsorted" \
  -REFERENCE_SEQUENCE \
  Homo_sapiens_assembly38.fasta
```

This produces the error:
```
java.lang.IllegalStateException: Aligned record iterator (NB500989:333:HKYJNAFX2:1:11101:10000:1915) is behind the unmapped reads (NB500989:333:HKYJNAFX2:1:11101:24447:1024)
	at picard.sam.AbstractAlignmentMerger.mergeAlignment(AbstractAlignmentMerger.java:557)
	at picard.sam.SamAlignmentMerger.mergeAlignment(SamAlignmentMerger.java:186)
	at picard.sam.MergeBamAlignment.doWork(MergeBamAlignment.java:368)
	at picard.cmdline.CommandLineProgram.instanceMain(CommandLineProgram.java:308)
	at org.broadinstitute.hellbender.cmdline.PicardCommandLineProgramExecutor.instanceMain(PicardCommandLineProgramExecutor.java:37)
	at org.broadinstitute.hellbender.Main.runCommandLineProgram(Main.java:160)
	at org.broadinstitute.hellbender.Main.mainEntry(Main.java:203)
	at org.broadinstitute.hellbender.Main.main(Main.java:289)
```

Despite the fact that option `-SORT_ORDER "unsorted"` is being used and the two BAM files have the reads in the same order:
```
$ samtools view unmapped.bam | cut -f1
NB500989:333:HKYJNAFX2:1:11101:24447:1024
NB500989:333:HKYJNAFX2:1:11101:24447:1024
NB500989:333:HKYJNAFX2:1:11101:10000:1915
NB500989:333:HKYJNAFX2:1:11101:10000:1915

$ samtools view aligned.unmerged.bam | cut -f1
NB500989:333:HKYJNAFX2:1:11101:24447:1024
NB500989:333:HKYJNAFX2:1:11101:24447:1024
NB500989:333:HKYJNAFX2:1:11101:10000:1915
NB500989:333:HKYJNAFX2:1:11101:10000:1915
```
Is there a way to run `MergeBamAlignment` letting the tool know that the order of the reads in the input BAMs is the same?

It seems then that the only workaround is to let `FastqToSam` sort the reads, which seems an unnecessary computational step since eventually the reads will be sorted by coordinate anyway.

This problem was observed when running together the WDL pipelines [paired-fastq-to-unmapped-bam.wdl](https://github.com/gatk-workflows/seq-format-conversion/blob/master/paired-fastq-to-unmapped-bam.wdl) and [processing-for-variant-discovery-gatk4.wdl](https://github.com/gatk-workflows/gatk4-data-processing/blob/master/processing-for-variant-discovery-gatk4.wdl) where the former runs `FastqToSam` and the latter runs `SamToFastq` and `MergeBamAlignment`.

**Comment by jamesemery** [comment]
Hello @freeseek. It should be noted that you might be somewhat confused about what the -SO arguments do in the tools:
`-SORT_ORDER "unsorted"` which has the help description `"The order in which the merged reads should be output.  Default value: coordinate. Possible values: {unsorted, queryname, coordinate, duplicate, unknown}."` For MergeBamAlignment that argument doesn't change how the input sort order is treated but rather is used to determine whether the tool will resort your reads before outputting them. Currently it looks like MergeBamAlignment requires queryname sorted inputs: `"Original SAM or BAM file of unmapped reads, which must be in queryname order.  Reads MUST be unmapped."` and will crash with the unhelpful message you encountered. It may be possible to loosen this restriction to queryname-grouped inputs (since it really primarily needs the mates adjacent to each other but not necessarily in readname order). The place to have this discussion would be here https://github.com/broadinstitute/picard since this tool is a picard tool that is called into by GATK and any changes would need to be approved by the maintainers of that repo.. Furthermore, I would recommend using the gatk forums here: https://gatk.broadinstitute.org/hc/en-us/community/topics There are users and moderators there who might be able to help you answer questions like this for you.

**Comment by freeseek** [comment]
@jamesemery thank you! My bad, since I was running `MergeBamAlignment` through the GATK release from this github repository, I mistakenly assumed this would be the correct place where to report an issue. You are right though that this is the intended behavior of `MergeBamAlignment`. I did not see this as a bug, but rather as an issue with the `MergeBamAlignment` design which inflexibly requires to sort reads in queryname order adding unnecessary computational complexity to alignment pipelines. In any case, I followed your advice and posted this issue [here](https://github.com/broadinstitute/picard/issues/1710). Thank you for your help.
